### DISCUSSION TITLE
Simple type-safe `@dataclass` CLI with `hydra_zen.zen()`

### DISCUSSION THREAD

**Discussion by kpe** [mit-ll-responsible-ai/hydra-zen#738] [Q&A] [answered]
(I'm very new to hydra and hydra-zen, and cannot quite wrap my head around it) 
I'm trying to create a simple CLI with `hydra_zen.zen()` (while staying as pythonic as possible) like:

```bash
python run.py task=train train.lr=0.01
python run.py task=eval eval.model_path=model.pth
```

I understand, I could do something like:

```python 
from dataclasses import dataclass, field
from hydra_zen import ZenStore, zen

@dataclass
class TrainConfig:
    lr: float = 0.01
    num_epochs: int = 5
    batch_size: int = 32

@dataclass
class EvalConfig:
    model_path: str = "model.pth"
    batch_size: int = 16

@dataclass
class MainConfig:
    task = "train"
    train: TrainConfig = field(default_factory=TrainConfig)
    eval: EvalConfig = field(default_factory=EvalConfig)

# def main(cfg: MainConfig) -> None:
#     print(OmegaConf.to_yaml(cfg))

def main_fn(task: str, train: TrainConfig, eval: EvalConfig) -> None:
    print(OmegaConf.to_yaml(cfg))

store = ZenStore()
store(MainConfig, name="main")

if __name__ == "__main__":
    store.add_to_hydra_store()
    zen(main_fn).hydra_main(
        config_name="main",
        config_path=None,
        version_base="1.3",
    )
```
to call `main_fn()` by defining the arguments using a `@dataclass` , but would it be somehow possible to call the `main` method directly, which has all the configuration passed as a single `cfg: MainConfig` argument, while keeping the above CLI interface?
If I would just use `zen(main)` then the configuration would be expected as `cfg.train.lr=0.01` and not `train.lr=0.01`. Is there some way around this?

**Comment by rsokl** [comment, marked as answer]
What you are asking for is Hydra's default behavior; just use `hydra.main` instead of `zen`. 

But note that it is typically better design to have your function deal not with configs but the objects that are instantiated from those configs. That is the main differentiating factor of hydra-zen apps vs Hydra apps.

**Reply by kpe** [reply]
I see, I guess the following part of the existing documentation explains the point - https://mit-ll-responsible-ai.github.io/hydra-zen/explanation/dont_repeat_yourself.html
